raysect/core/math/function/float/function2d/interpolate/interpolator2darray.pyx

```diff
@@ -0,0 +1,1098 @@
+
+    cdef double evaluate(self, double px, double py, int index_x, int index_y) except? -1e999:
+        """
+        Calculates interpolated value at given point. 
+
+        :param double px: the point for which an interpolated value is required.
+        :param double py: the point for which an interpolated value is required.
+        :param int index_x: the lower index of the bin containing point px. (Result of bisection search).   
+        :param int index_y: the lower index of the bin containing point py. (Result of bisection search).   
+        """
+        raise NotImplementedError('_Interpolator is an abstract base class.')
+
+    cdef double analytic_gradient(self, double px, double py, int index_x, int index_y, int order_x, int order_y):
+        raise NotImplementedError('_Interpolator is an abstract base class.')
+
+
+cdef class _Interpolator2DLinear(_Interpolator2D):
+    """
+    Linear interpolation of 2D function.
+
+    :param x: 1D memory view of the spline point x positions.
+    :param y: 1D memory view of the spline point y positions.
+    :param f: 2D memory view of the function value at spline point x, y positions.
+    """
+
+    ID = 'linear'
+
+    def __init__(self, double[::1] x, double[::1] y, double[:, ::1] f):
+        super().__init__(x, y, f)
+
+    cdef double evaluate(self, double px, double py, int index_x, int index_y) except? -1e999:
+        return linear2d(
+            self._x[index_x], self._x[index_x + 1], self._y[index_y], self._y[index_y + 1],
+            self._f[index_x:index_x + 2, index_y:index_y + 2], px, py
+        )
+
+    cdef double analytic_gradient(self, double px, double py, int index_x, int index_y, int order_x, int order_y):
+        """
+        Calculate the normalised derivative of specified order in a unit square.
+        
+        The order of the derivative corresponds to order_x and order_y as the number of times differentiated. For 
+        example order_x = 1 and order_y = 1 is d2f/dxdy. The normalised gradient is calculated of the bilinear 
+        function f(x, y) = a0 + a1x + a2y + a3xy, which is the product of 2 linear functions. The derivatives are 
+        therefore df/dx = a1 + a3y ; df/dy = a1 + a3x ; d2f/dxdy = a3. The derivatives are calculated on the normalised 
+        unit square.
+
+        :param double px: the point for which an interpolated value is required.
+        :param double py: the point for which an interpolated value is required.
+        :param int index_x: the lower index of the bin containing point px. (Result of bisection search).   
+        :param int index_y: the lower index of the bin containing point py. (Result of bisection search).   
+        :param int order_x: the derivative order in the x direction.
+        :param int order_y: the derivative order in the y direction.
+        """
+        cdef double df_dn
+        cdef double[4] a
+
+        self.calculate_coefficients(index_x, index_y, a)
```

Comment:
It would be more computationally efficient to simply write the values. Inserting an if statement into a calculation path where it is not necessary is bad re. optimisation. Modern CPUs have long computation pipelines that have to be flushed if a branch prediction fails, hitting performance severely. The code here could be entirely deterministic avoiding the risk of a branch prediction failure. The calculation should be moved out and if necessary repeated as many times as needed so the compiled code ends up efficient.

Reply:
...or just make 4 separate functions without the conditional test.

Reply:
I have now changed the bi and tri linear coefficient calculations to separate functions for each coefficient (4 and 8 functions respectively).